File: 290-word-pattern/word-pattern.java
import java.util.HashMap;

class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] arr = s.split(" ");
        if(pattern.length() != arr.length) {
            return false;
        }   
        HashMap<Character, String> hm = new HashMap<>();
        
        for(int i = 0; i < pattern.length(); i++) {
            char ch = pattern.charAt(i);
        
            if(hm.containsValue(arr[i]) && !hm.containsKey(ch)) {
                return false;
            }
    
            if(hm.containsKey(ch) && !hm.get(ch).equals(arr[i])) {
                return false;
            }
            
      
            hm.put(ch, arr[i]);
        }
        
        return true;
    }
}

File: 290-word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()

        # Number of pattern characters and words must be same
        if len(pattern) != len(words):
            return False

        mapPW = {}  # Pattern -> Word
        mapWP = {}  # Word -> Pattern

        for i in range(len(pattern)):

            p = pattern[i]
            w = words[i]

            # Check Pattern -> Word mapping
            if p in mapPW:
                if mapPW[p] != w:
                    return False
            else:
                mapPW[p] = w

            # Check Word -> Pattern mapping
            if w in mapWP:
                if mapWP[w] != p:
                    return False
            else:
                mapWP[w] = p

        return True